[WIP] Models and more

samselikoff opened the pull request:
- [x] basic schema, model, relation
- [x] belongs to
- [x] has many
- [x] update route handler api - probably inject `schema` instead of `db` (`db` accessible via `schema.db`)

seawatts commented:
Hey do you think it might make sense to break this into smaller PRs? This one is getting huge.

samselikoff commented:
Yah, this is the first time I've really worked on such a big feature in a separate branch like this. What would be the way to split it up? All the new objects are related to each other..

seawatts commented:
Yeah thats always hard. What I do sometimes just to make it easier is pull some of the code onto separate branches and then merge back into this branch just to make it a little easier to review. But again it's kinda hard when everything becomes very coupled up.

seawatts commented on CHANGELOG.md:
```diff
@@ -1,5 +1,12 @@
 
+## next
+
+Changes:
```
We may want to add a deprecation warning for this in v0.0.30?
And then actually do the release in v0.1.x

seawatts commented:
Maybe you could separate it by your checkpoints above?
belongsTo,
hasMany,
etc

samselikoff commented:
i would be dead without tests.

also, es6 works under /addon now!!111 time to clean up these classes

samselikoff commented:
@seawatts I'm happy to rebase this, how should I organize? Put belongsTo + tests in one commit, hasMany + tests in another?

seawatts commented:
@samselikoff Yeah that sounds pretty good to me. It will just make it easier to read the PR and find potential bugs.

samselikoff commented:
I'm going to merge this soon, bc I should be able to undo the two breaking changes, now that I'm dropping ie8 support and can use getter/setters:

 - [ ] I changed db.collection to db.collection.all(). Will revert
 - [ ] I changed db.collection.update(target, attrs) to db.collection.update(attrs, target). Will revert

I want to do this so I can go ahead and merge all this in, as it's completely backwards compatible. This way we can work on serializers in a new PR.

seawatts commented:
That sounds great @samselikoff! :+1: Thanks for continuing to work on this! I can't wait for this to land!

samselikoff commented:
:sweat: whew!

ok so these are usable now, but not really useful. next step will be to add base serializers - essentially, if a route handler returns a model, just call `model.attrs`.

seawatts commented:
:+1:

cibernox commented:
:confetti_ball:

bdvholmes commented:
:tada:

jamesdixon commented:
@samselikoff is there any documentation on the associations you implemented in this PR?

samselikoff commented:
no, bc they really need serializers to be effective. well kind of. but yah I want to implement those first (getting very close!)

jamesdixon commented:
Gotcha, thanks!



From: Sam Selikoff <notifications@github.com>
Reply: samselikoff/ember-cli-mirage <reply@reply.github.com>>
Date: August 12, 2015 at 9:51:14 PM
To: samselikoff/ember-cli-mirage <ember-cli-mirage@noreply.github.com>>
Cc: jamesdixon <jim.w.dixon@gmail.com>>
Subject:  Re: [ember-cli-mirage] [WIP] Models and more (#82)  

no, bc they really need serializers to be effective. well kind of. but yah I want to implement those first (getting very close!)

—
Reply to this email directly or view it on GitHub.

dslh commented:
Hi, just a quick bump for docs on this feature, are there any available? The official docs at ember-cli-mirage.com are still [waiting for this PR to land](http://www.ember-cli-mirage.com/docs/v0.1.x/seeding-your-database/).

I'd be happy to contribute a nicely-written patch to the docs myself if that's possible, if I could just get a quick usage example to get me started.

samselikoff commented:
@dslh I've kept the default docs pointed to the latest release version, but in the bottom-left you can select the {beta version of the docs](http://www.ember-cli-mirage.com/docs/v0.2.0-beta.3/). You'll see the docs have been restructured. Have a look at the [Defining routes](http://www.ember-cli-mirage.com/docs/v0.2.0-beta.3/defining-routes/) guide for an overview on models, and also the Models, Schema and Serializer reference docs for more info.

dslh commented:
Awesome! Thanks so much for the fast reply :smile:
